### ISSUE TITLE
Lint on useless transmutes

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#441]
https://github.com/rust-lang/rust/issues/11400

We should lint on transmutes between the same type, e.g. `transmute::<u8,u8>(...)`

The easiest thing to do here is to find `ExprCall`s which resolve to `transmute`, and checking if the `expr_ty` of the argument matches that of the whole call expression.

**Comment by llogiq** [comment]
This absolutely needs a macro check. And it should be `mem::transmute`, right? Just to make sure we don't catch anyone's own `transmute(_)` function, say, in a medieval fantasy game...

**Comment by Manishearth** [comment]
Well, yeah, match_def_path takes care of that

**Comment by apasel422** [comment]
I'm trying to implement this, but I'm wondering how to get a `DefId` from the `Path` in the `ExprCall` so that I can call `match_def_path`.

**Comment by Manishearth** [comment]
use the `def_map` in `tcx` on the `NodeId` of the call?

**Comment by apasel422** [comment]
Thanks! And to check if two types are equal?

**Comment by Manishearth** [comment]
`ty::Ty` has an `Eq` impl.

**Comment by Manishearth** [comment]
I can't recall if that actually _works_, or if it computes some arcane form of equality :stuck_out_tongue_closed_eyes:. Worth a try though. If it works, see if it can also catch:

```
fn foo<T>(t: T) {
transmute::<T,T>(t);
}
```

or something.

**Comment by Manishearth** [comment]
It looks like it does a pointer comparison, which sound scary but since it's in an explicit impl I'll believe it. Probably because all the TySs are borrowed from the same cache/arena.

**Comment by llogiq** [comment]
Otherwise `eq_op.rs` should have a comparison function.
